m4opt/models/extinction/airmass.py

```diff
@@ -0,0 +1,338 @@
+
+
+class Airmass:
+    """
+    Airmass:
+
+    Object to calculate airmass at a given observatory location,
+    for a provided target SkyCoord.
+
+    Airmass class currently supports two models:
+    1. 'simple' : Plane-Parallel atmosphere, airmass = 1/cos(zenith)
+    2. 'kastenyoung' : Kasten+Young89 interpolation fit to data[1]_
+
+    References
+    ----------
+    .. [1] Kasten, F.; Young, A. T. (1989). "Revised optical air mass
+           tables and approximation formula". Applied Optics. 28 (22):
+           4735-4738. doi:10.1364/AO.28.004735. PMID 20555942.
+
+    Examples
+    --------
+    The Airmass object requires a ground observer location at initialization,
+    passed in via `astropy.coordinates.EarthLocation`:
+
+    >>> from m4opt.models.extinction import Airmass
+    >>> import astropy.units as u
+    >>> from astropy.coordinates import SkyCoord, EarthLocation
+    >>> from astropy.time import Time
+
+    >>> place_airm = Airmass(EarthLocation(lat=41.3*u.deg, lon=-74*u.deg,\
+                             height=390*u.m))
+
+    or choose a pre-defined location, e.g. Kitt Peak:
+
+    >>> kpno_airm = Airmass.kpno()
+
+    To evaluate an airmass for an observing target, we pass in the SkyCoord
+    and the Time for the target, using `at()`:
+
+    >>> time = Time('2012-7-13 07:00:00')
+    >>> target = SkyCoord.from_name('m33')
+    >>> place_airm.at(target, time)
+    <Quantity 1.53952747>
+
+    The default airmass model is a plane-parallel atmosphere, which defines
+    the airmass as the secant of the zenith angle. We can choose a different
+    model if we wish to use more realistic values, especially near
+    the horizon:
+
+    >>> time = Time('2012-7-13 03:00:00')
+    >>> place_airm.at(target, time)
+    <Quantity 442.57856613>
+    >>> place_airm.set_model("kastenyoung")
+    >>> place_airm.at(target,time)
+    <Quantity 36.05249807>
+    """
+
+    def __init__(self, earth_location, airmass_model=None):
+        if not isinstance(earth_location, EarthLocation):
+            raise TypeError("Input earth_location must be of type \
+                             astropy.coordinates.earth.EarthLocation")
+
+        self.earth_loc = earth_location
+        self.set_model(airmass_model)
+        return
+
+    def set_model(self, model_name):
+        if model_name is None:
+            self.model = simple_airmass
+        elif model_name.lower() not in airmass_models.keys():
+            raise AttributeError("model must be one of {0}".format(
+                                 airmass_models.keys()))
+        else:
+            self.model = airmass_models[model_name.lower()]
+
+    def at(self, target_coord, obs_time):
+        """
+        returns airmass at target sky location at obs_time
+        """
+        if not isinstance(target_coord, SkyCoord):
+            raise TypeError("argument target_coord must be of type \
+                            astropy.coordinates.sky_coordinate.SkyCoord")
```

Comment:
I don't think this should check explicitly for `SkyCoord`, because this machinery should also work for e.g. `AltAz` or `ICRS` objects.  Instead I suggest duck-typing, something like 
```suggestion
        if not hasattr(target_coord, 'transform_to'):
            raise TypeError("argument target_coord must be an astropy coordinates object")
```